### DISCUSSION TITLE
14. HTTP 무상태성은 무엇인가요? ‘오늘 하루 그만보기’ 같은 기능은 어떻게 구현하나요?

### DISCUSSION THREAD

**Discussion by eun-byeol** [jewelry-box/cs-interview#18] [General]
HTTP 무상태성은 무엇인가요? ‘오늘 하루 그만보기’ 같은 기능은 어떻게 구현하나요?

**Comment by jongmee** [comment]
- 서로 다른 요청 간에 상태를 유지하고 있지 않는게 무상태성이다. 이전 요청의 상태를 이후 요청이 알 수 없다.
- '오늘 하루 그만보기' 기능은 브라우저 저장소를 활용할 수 있다. 쿠키 혹은 로컬 스토리지를 활용한다.
- 쿠키는 서버에 요청할 때마다 함께 전송되면 불필요한 네트워크 트래픽이 발생할 수 있기 때문에 적절한 `path`를 설정한다.

**Comment by cutehumanS2** [comment]
- **HTTP의 무상태성**은 서버가 클라이언트의 요청을 독립적인 요청으로 처리하며, 이전 요청 정보를 기억하지 않는 특성입니다.
    - 예를 들어, 로그인 후에도 매 요청마다 인증 정보를 포함해야 하는 것이 무상태성의 특징입니다.
- ‘오늘 하루 그만보기’ 같은 기능은 **쿠키** 또는 **로컬 스토리지** 등을 사용하여 클라이언트 또는 서버 측에서 상태를 관리해야 합니다.

<br>

**[‘오늘 하루 그만 보기’를 쿠키와 로컬 스토리지 중 어느 방식으로 구현하는 것이 좋을까요?]** 
- 일반적으로 팝업 같은 단순한 상태는 로컬 스토리지로 구현하는 게 적합합니다.
- 로그인 유지 등 서버 상태와 연관된 경우에는 쿠키가 적합합니다.

**Comment by linirini** [comment]
### 쿠키 vs 로컬 스토리지
항목 | 쿠키 (Cookie) | 로컬 스토리지 (Local Storage)
-- | -- | --
데이터 전송 | 서버에 자동 전송됨 (HTTP 요청 시 포함) | 브라우저 내에서만 접근 가능 (서버 전송 ❌)
XSS 취약점 | HttpOnly 설정 가능 → 보호 가능 | JavaScript에서 접근 가능 → XSS 취약
CSRF 취약점 | 자동 전송되므로 CSRF 위험 있음 | 서버로 자동 전송되지 않아 CSRF 안전
보안 옵션 | HttpOnly, Secure, SameSite 설정 가능 | 보안 옵션 부족 (별도 조치 필요)
저장 가능 크기 | 4KB 제한 | 5MB 이상 가능
속도 | 서버 요청 시마다 포함 → 네트워크 비용 발생 | 네트워크 비용 없음 (빠름)
사용 예시 | 세션 관리, 인증 토큰 저장 (JWT, Session ID) | API 응답 캐싱, 사용자 설정 저장
